src/net.cpp

```diff
@@ -1608,6 +1608,43 @@ std::unordered_set<Network> CConnman::GetReachableEmptyNetworks() const
     return networks;
 }
 
+int CConnman::GetFullOutboundAndManualCount(Network net) const
+{
+    LOCK(m_nodes_mutex);
+    int outbound_peers{0};
+    std::optional<Network> net_alt;
+    // Treat clearnet (IPv4 and IPv6) as one network
+    if (net == Network::NET_IPV4) net_alt = Network::NET_IPV6;
+    if (net == Network::NET_IPV6) net_alt = Network::NET_IPV4;
+    for (const auto& pnode : m_nodes) {
+        if (pnode->fSuccessfullyConnected && !pnode->fDisconnect && (pnode->IsFullOutboundConn() || pnode->IsManualConn())) {
+            auto net_peer{pnode->addr.GetNetwork()};
+            if (net == net_peer || (net_alt.has_value() && net_alt == net_peer)) {
+                ++outbound_peers;
+            }
+        }
+    }
+    return outbound_peers;
+}
+
+bool CConnman::MaybePickPreferredNetwork(std::optional<Network>& network)
+{
+    std::vector<Network> preferred_networks;
+    for (int n = 0; n < NET_MAX; n++) {
+        enum Network net = (enum Network)n;
+        if (net == NET_UNROUTABLE || net == NET_INTERNAL) continue;
+        if (IsReachable(net) && GetFullOutboundAndManualCount(net) == 0 && addrman.Size(net) != 0) {
+            preferred_networks.push_back(net);
+        }
+    }
+    if (preferred_networks.size() > 0) {
```

Comment:
There is no need to collect more than one element in `preferred_networks`, the previous loop can be stopped as soon as a match is found and the shuffling can be done on the order in which networks are iterated (untested):

```cpp
std::optional<Network> CConnman::MaybePickPreferredNetwork()
{
    std::array<Network, 5> nets{NET_IPV4, NET_IPV6, NET_ONION, NET_I2P, NET_CJDNS};
    Shuffle(nets.begin(), nets.end(), FastRandomContext());
    auto peers_per_net = GetFullOutboundAndManualCounts();
    for (const auto net : nets) {
        if (IsReachable(net) && peers_per_net[net] == 0 && addrman.Size(net) != 0) {
            return net;
        }
    }
    return std::nullopt;
}
```